lib/web_ui/lib/src/engine/configuration.dart

```diff
@@ -18,47 +25,101 @@
+/// environment variables are provided, the `canvasKitBaseUrl` entry is used.
 
 @JS()
 library configuration;
 
 import 'package:js/js.dart';
+import 'package:meta/meta.dart';
+import 'dom.dart';
 
 /// The version of CanvasKit used by the web engine by default.
 // DO NOT EDIT THE NEXT LINE OF CODE MANUALLY
 // See `lib/web_ui/README.md` for how to roll CanvasKit to a new version.
 const String _canvaskitVersion = '0.37.0';
 
 /// The Web Engine configuration for the current application.
-FlutterConfiguration get configuration => _configuration ??= FlutterConfiguration(_jsConfiguration);
+FlutterConfiguration get configuration => _configuration ??= FlutterConfiguration.fromJsGlobals(_jsConfiguration);
 FlutterConfiguration? _configuration;
 
 /// Sets the given configuration as the current one.
 ///
 /// This must be called before the engine is initialized. Calling it after the
 /// engine is initialized will result in some of the properties not taking
 /// effect because they are consumed during initialization.
+@visibleForTesting
 void debugSetConfiguration(FlutterConfiguration configuration) {
   _configuration = configuration;
 }
 
 /// Supplies Web Engine configuration properties.
 class FlutterConfiguration {
-  /// Constructs a configuration from a JavaScript object containing
-  /// runtime-supplied properties.
-  FlutterConfiguration(this._js);
+  /// Constructs an unitialized configuration object.
+  @visibleForTesting
+  FlutterConfiguration();
 
-  final JsFlutterConfiguration? _js;
+  /// Constucts a "tainted by JS globals" configuration object.
+  ///
+  /// This is deprecated, and warns the user about the new API.
+  FlutterConfiguration.fromJsGlobals(JsFlutterConfiguration? config) {
+    if (config != null) {
+      domWindow.console.warn('window.flutterConfiguration is now deprecated.\n'
+      'Use engineInitializer.initializeEngine(config) instead.\n'
+      'See: https://docs.flutter.dev/development/platform-integration/web/initialization');
+      _createdFromJsGlobals = true;
+      _configuration = config;
+    }
+    if (_requestedRendererType != null) {
+      domWindow.console.warn('window.flutterWebRenderer is now deprecated.\n'
+      'Use engineInitializer.initializeEngine(config) instead.\n'
+      'See: https://docs.flutter.dev/development/platform-integration/web/initialization');
+    }
+  }
+
+  bool _createdFromJsGlobals = false;
+  JsFlutterConfiguration? _configuration;
+
+  /// Sets a value for [_configuration].
+  ///
+  /// This method is called by the engine initialization process, through the
+  /// [initEngineServices] method.
+  ///
+  /// This method throws an AssertionError, if the _configuration object has
+  /// been set to anything non-null through the [FlutterConfiguration.fromJsGlobals]
+  /// constructor.
+  void setUserConfiguration(JsFlutterConfiguration? configuration) {
+    if (configuration != null) {
+      assert(!_createdFromJsGlobals, 'Do not mix-and-match configuration styles. '
```

Comment:
It's unclear what "mix-and-match" means. I'd say "Using both the legacy window.flutterConfiguration and initialization configuration simultaneously is not supported."

Reply:
Final text:

"Use engineInitializer.initializeEngine(config) only. Using the (deprecated) window.flutterConfiguration and initializeEngine configuration simultaneously is not supported."